File: JavaSelenium/src/test/java/com/taskmanager/e2e/pages/LoginPopup.java
package com.taskmanager.e2e.pages;

import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import com.taskmanager.e2e.models.User;

public class LoginPopup extends BasePage {

  private static final String TITLE_TEST_ID = "login-title";
  private static final String USERNAME_INPUT_TEST_ID = "username-input";
  private static final String PASSWORD_INPUT_TEST_ID = "password-input";
  private static final String LOGIN_BUTTON_TEST_ID = "login-submit";

  public LoginPopup(WebDriver driver) {
    super(driver);
  }

  private WebElement waitForOpen() {
    return waitVisible(Locators.testId(TITLE_TEST_ID));
  }

  private WebElement usernameInput() {
    return driver.findElement(Locators.testId(USERNAME_INPUT_TEST_ID));
  }

  private WebElement passwordInput() {
    return driver.findElement(Locators.testId(PASSWORD_INPUT_TEST_ID));
  }

  private WebElement loginButton() {
    return driver.findElement(Locators.testId(LOGIN_BUTTON_TEST_ID));
  }

  public void login(User user) {
    waitForOpen();
    usernameInput().sendKeys(user.name());
    passwordInput().sendKeys(user.password());
    loginButton().click();
  }
}

File: APIandUItesting/apps/web-e2e/pages/login_popup.ts
import { LogAction } from '../annotations/log-action';
import { User } from '../tests/types';
import { BasePage } from './base_page';

const LOGIN_POPUP_TEST_IDS = {
  title: 'login-title',
  usernameLabel: 'username-label',
  usernameInput: 'username-input',
  passwordLabel: 'password-label',
  passwordInput: 'password-input',
  loginButton: 'login-submit',
  forgotPasswordButton: 'forgot-password-button',
  registerButton: 'register-button',
  loginError: 'login-error',
};

export class LoginPopup extends BasePage {
  title = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.title);
  usernameLabel = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.usernameLabel);
  usernameInput = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.usernameInput);
  passwordLabel = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.passwordLabel);
  passwordInput = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.passwordInput);
  loginButton = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.loginButton);
  forgotPasswordButton = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.forgotPasswordButton);
  registerButton = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.registerButton);
  loginError = this.page.getByTestId(LOGIN_POPUP_TEST_IDS.loginError);

  @LogAction
  async login(user: User): Promise<void> {
    await this.usernameInput.fill(user?.name);
    await this.passwordInput.fill(user?.password);
    await this.loginButton.click();
  }
}
